New --advertise option to specify different bind/accept IPs

noonio opened the pull request:
This is relevant when needing to advertise an IP address for other (external) parties to connect to, if it is not something you yourself can bind to (because it's a public IP address, say.)

Todo:

- [x] Docs
- [x] Tests
- [x] Update all examples to use `--listen` instead of `--host` and `--port`

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Draft a new --advertise option to separate listening from advertising public IPs

noonio committed:
Changelog; docs; tests

v0d1ch commented on CHANGELOG.md:
```diff
@@ -12,7 +12,9 @@ changes.
-  - The same network options to `hydra-node` required (`--host`, `--port` and `--peer`) and need to match across the Hydra network.
+  - The peer network options to `hydra-node` (`--peer`) need to match across the Hydra network.
+  - `--host` and `--port` have been removed and we now have `--listen` and `--advertise`.
+  - `--listen` can be the same as the hold `--host/--port` combined, but you may require `--advertise` to set your _public_ IP address and port combination, if you cannot bind to that on the local machine.
```
```suggestion
  - `--listen` can be the same as the old `--host/--port` combined, but you may require `--advertise` to set your _public_ IP address and port combination, if you cannot bind to that on the local machine.
```

v0d1ch reviewed (approved):
Nice! Found one typo

noonio committed:
Update CHANGELOG.md

Co-authored-by: Sasha Bogicevic <Sasha.Bogicevic@iohk.io>

ffakenz commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -355,8 +355,8 @@ withHydraNode tracer chainConfig workDir hydraNodeId hydraSKey hydraVKeys allNod
-                , host = "0.0.0.0"
-                , port = fromIntegral $ 5_000 + hydraNodeId
+                , listen = Host "0.0.0.0" (fromIntegral $ 5_000 + hydraNodeId)
+                , advertise = Nothing
```
Is this still opening a head with a cluster of peers?

noonio replied:
Nothing has changed from what it did before; i.e. I just moved the host/port combination into `listen`.

noonio committed:
More argument invocation fixups

noonio committed:
Crimes against IPv6 parsing; 127.0.0.1 instead of localhost

Not sure why the localhost -> 127.0.0.1 fix fixed things; but there you
go.

I commented out IPv6 tests for now as it should read them fine, but
of course requires a whole bunch of busywork to do that properly.

Minor crimes committed against `readHost` to just look for the last `:`
instead of the first one.

noonio committed:
Default is 0.0.0.0 not localhost

noonio committed:
Update schema
